cadquery/occ_impl/geom.py

```diff
@@ -167,8 +167,23 @@ def __init__(self, matrix=None):
 
         if matrix is None:
             self.wrapped = gp_Trsf()
-        else:
+        elif isinstance(matrix, gp_Trsf):
             self.wrapped = matrix
+        elif isinstance(matrix, list):
+            self.wrapped = gp_Trsf()
+            if len(matrix) == 12:
+                self.wrapped.SetValues(*matrix)
+            elif len(matrix) == 16:
+                # Only use first 12 values - the last 4 must be [0, 0, 0, 1].
+                last4 = matrix[12:]
+                if last4 != [0., 0., 0., 1.]:
+                    raise ValueError("Expected the last 4 values to be [0,0,0,1], but got: {}".format(last4))
```

Comment:
Technically this isn't always true.
https://en.wikipedia.org/wiki/Scaling_(geometry)#Using_homogeneous_coordinates

personally I'd leave this check out; if an end user is explicitly defining affine transform matrices, then validity is arguably scoped in their code, not ours.

Also, it's technically _just_ a 4x4 matrix, it's not necessarily exclusively for the purposes of a transform (although unlikely).
But I do think the assumption given 12 elements is fair enough.

I'll get off my soapbox now, what do you think?

Reply:
Thanks for the feedback!

You're definitely right that in the general case, the last row isn't required to be `[0,0,0,1]`. Because the `Matrix` class is backed by OCC's `gp_Trsf` though, it is restricted to transform matrices and not arbitrary 4x4s.

The `gp_Trsf` `SetValues()` method only lets you set the first 3 rows and assumes the last to be `[0,0,0,1]` because it stores a "normalized" matrix.

It would be possible to assume the last row given as input is a more general `[0,0,0,x]`, then normalize the first three rows depending on the value of `x`. Do you think this would be better than just asserting that it's the expected `[0,0,0,1]`?

See the `SetValues()` docs here: https://www.opencascade.com/doc/occt-6.9.1/refman/html/classgp___trsf.html#a236867197c4fc79c57b5e34cf643005f.

Reply:
> Because the Matrix class is backed by OCC's `gp_Trsf` though, it is restricted to transform matrices and not arbitrary 4x4s

🤦‍♂️  you're absolutely right.
If I just read the next line properly 🙄 

> It would be possible to assume the last row given as input is a more general `[0,0,0,x]`, then normalise the first three rows depending on the value of `x`.

No, that's scope-creep.
If somebody posts an issue with a legitimate use case for this in the future, _then_ it could be added... but for now, I'm doing a full 180°, your original implementation is good. 👍

Reply:
Following the OCC convention (so gp_Trsf) makes sense here. NB: this is a low level "*_impl" object. "Normal" users won't be in principle touching it.